### ISSUE TITLE
tangent line to the compound curve

### ISSUE THREAD

**Issue by supipd** [jsxgraph/jsxgraph#318]
I have a problem with the tangent line to the compound curve (functiongraph). At the breaking point, the tangent jumps as shown on picture.  Any idea how to solve? 
Maybe derivation from both sides, to overcame precision / limit problem ?
Ideally, I would need a function to create tangent line to the curve, passing through the specified point outside of curve.

![tangent_of_complex_functiongraph](https://user-images.githubusercontent.com/1335871/82901961-fe468c80-9f5e-11ea-854e-0e9b58d31f12.PNG)

Very good library, thanks.

**Comment by alfredwassermann** [comment]
These are two questions.

1) Tangent to the compound curve: I'm not sure what you expect here. Do you want the tangent to disappear at the points where the derivative does not exist? 
2) Create tangent line to the curve, passing through the specified point outside of curve: 
The following code is an example: 

```
    var p = board.create('point', [4, 2], {name: 'p'});
    var f = (x) => 3*x*x + 1*x  + 1;
    var df = JXG.Math.Numerics.D(f);
    var plotf= board.create('functiongraph', [f]);
    var g = (x) => (x - p.X()) * df(x) + p.Y() - f(x);

    var q = board.create('point', [
        	function() { 
  	        var x = JXG.Math.Numerics.root(g, p.X());
  	        return [x, f(x)]; }
        ], {name: 'q'});

    board.create('line', [p, q], {strokeWidth: 1, dash: 2});
```
See it live at  https://jsfiddle.net/4vetmc9r/1/ .
But beware, the root finding function will find one root. If there are multiple solution, there might be problems.

If you want to continue the discussion, do you mind if we switch to stackoverflow. Here on git hub we would lik to restrict the discussion to bugs or feature requests of the code.

**Comment by supipd** [comment]
There are tangent from right and tangent from left near the breaking point, and the their angles are very different. The normal line to the half of their's angle can be the tangent at the breaking point. Your're right, derivate in breaking point is not defined in theory ... but in practice (and precision possibilities), "special" tangent line can  be intuitive and usefull. (Breaking point between segments can be miniature curve from one slope to another, in reality).

Thank for provided code, is very usefull, I've used it instead of moving gliders and theirs tanget. 
But results are not entirely satisfactory. For finding appropriate roots at several breaking points must be futher calculations, to obtain (intuitively clear) solution.
Can be seen it in demorun 
http://www.sbmintegral.sk/TUZV/apps/v5_202005191901/html/fragment5.xhtml?demorun=9

I'm not very good at mathematical theory, so I'm probably not enough for a deeper discussion on stackoverflow.

**Comment by alfredwassermann** [comment]
Indeed, in the case that the curve consists of two line segments, `JXG.Math.Numerics.D` will determine the slope of the tangent in the join exactly as you intend with your "special" tangent line. The reason is that `JXG.Math.Numerics.D` uses the symmetric difference quotient to approximate the derivative:

    f(x+h) - f(x-h) / (2h)

Here is an example (see it live at https://jsfiddle.net/fLthpb3x/)

    var f = function(x) {
    	if (x <= 0) {
      	return 3;
      }
      if (x <= 1) {
  	    return -2 * x + 3;
      }
      return 1;
    }

    var plot = board.create('functiongraph', f);
    var glide = board.create('glider', [1, 1, plot]);
    var tang = board.create('tangent', [glide]);

For the other problem, computing the tangent from an outside point, one suggestion is to use an interval for the root finding algorithm - in case you know lower and upper bounds for the tangent point:

    var q = board.create('point', [
	function() { 
  	var x = JXG.Math.Numerics.root(g, [-10, p.X()]);
  	return [x, f(x)]; }
    ], {name: 'q'});

See it at https://jsfiddle.net/q70obst9/, the red curve is the function graph of the function for which we search the root.

**Comment by alfredwassermann** [comment]
By the way, your application is fascinating (but the video is very fast if one does not know what it is about).

**Comment by supipd** [comment]
It's a application, which uses Glaser method to calculate the vapor condensation zones, vapor diffusion, and amount condensate at various temperature and humidity conditions, in various walls of buildings. Method is based on finding tangents to function of saturated vapor pressure (complicated, partially exponential function, depends on composition layers).
Because of different national standarts, app has only Slovak database, so there was no need to internationalize here.

Thanks for notice of "special" tangent, I'll try to dig in source code and find similar solution for my compound functions.
( Because of practically arbitrary user inputs, it will be hard define  precisions of numerical methods )

And, yes, I've used intervals for finding roots (and assignment to near segment's border), but in the presented TEST case (-20 degrees), it's seems, that it must be used some iterative method to specify intervals for finding all roots.

Btw, no video, DEMORUN is repeater of user events  (clicks, changes) in fully functional app. :-)) Can be stop,step,run,fast-forward in semi-opacitized panel at left top corner of screen. (So called "tracer" code I'll maybe send to Github in near future) 

Thanks again.
